src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal class RemoteJSDataStream : Stream
+    {
+        // Concerns with static `Instances`? Malicious actor could hijack another user's
+        // stream by (improbably) guessing the GUID.
+        // Maybe we put this in the JSRuntime for curcuit isolation? Would increate overhead as
+        // we'd have to go through CircuitHost, DotNetDispatcher and so on.
+        private readonly static Dictionary<Guid, RemoteJSDataStream> Instances = new();
+        private readonly JSRuntime _runtime;
+        private readonly Guid _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _cancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static Task SupplyData(string streamId, ReadOnlySequence<byte> chunk, string error)
+        {
+            if (!Instances.TryGetValue(Guid.Parse(streamId), out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We return silently as we still expect a couple of SupplyData calls after disposal IFF the RemoteJSDataStream was cancelled.
+                return Task.CompletedTask;
+            }
```

Comment:
Currently, the cancellation mechanism is based on sending a `Blazor._internal.cancelJSDataStream` call on disposal. Alternatively, we can update this to return a bool indicating whether or not the stream has been disposed. Thoughts/preferences?

Reply:
Think the second option is a bit simpler, and requires one fewer endpoint. Implemented in 46f4ddfca8.